frontend/src/types/schemaSystem.ts

```diff
@@ -0,0 +1,390 @@
+  value?: string;
+}
+
+export interface RulePayload {
+  [key: string]: StringPayload | StringArrayPayload | TemplatePayload;
+}
+
+export type SchemaRuleEngineType = "MYSQL" | "COMMON";
+
+export type CategoryType = "ENGINE" | "NAMING" | "QUERY" | "TABLE" | "COLUMN";
+
+export type RuleId =
+  | "engine.mysql.use-innodb"
+  | "table.require-pk"
+  | "naming.table"
+  | "naming.column"
+  | "naming.index.pk"
+  | "naming.index.uk"
+  | "naming.index.idx"
+  | "column.required"
+  | "column.no-null"
+  | "query.select.no-select-all"
+  | "query.where.require"
+  | "query.where.no-leading-wildcard-like";
+
+export interface RuleTemplate {
+  id: RuleId;
+  category: CategoryType;
+  engine: SchemaRuleEngineType;
+  description: string;
+  payload?: RulePayload;
+  level: RuleLevel;
+}
+
+interface BaseSchemaRuleType {
+  id: RuleId;
+  level: RuleLevel;
+}
+
+interface NamingFormatRuleType extends BaseSchemaRuleType {
+  format: string;
+}
+
+interface RequiredColumnRuleType extends BaseSchemaRuleType {
+  columnList: string[];
+}
+
+type SchemaPolicyRule =
+  | BaseSchemaRuleType
+  | NamingFormatRuleType
+  | RequiredColumnRuleType;
+
+export interface DatabaseSchemaReviewPolicy {
+  id: SchemaReviewPolicyId;
+
+  // Standard fields
+  creator: Principal;
+  createdTs: number;
+  updater: Principal;
+  updatedTs: number;
+
+  // Domain specific fields
+  name: string;
+  ruleList: SchemaPolicyRule[];
+  environmentIdList: number[];
+}
+
+export interface DatabaseSchemaReviewPolicyCreate {
+  // Domain specific fields
+  name: string;
+  ruleList: SchemaPolicyRule[];
+  environmentIdList: number[];
+}
+
+export type DatabaseSchemaReviewPolicyPatch = {
+  // Domain specific fields
+  name?: string;
+  ruleList?: SchemaPolicyRule[];
+  environmentIdList?: number[];
+};
+
+interface RuleCategory {
+  id: CategoryType;
+  ruleList: RuleTemplate[];
+}
+
+export interface SchemaReviewTemplate {
+  name: string;
+  imagePath: string;
+  ruleList: RuleTemplate[];
+}
+
+const categoryOrder: Map<CategoryType, number> = new Map([
+  ["ENGINE", 5],
+  ["NAMING", 4],
+  ["QUERY", 3],
+  ["TABLE", 2],
+  ["COLUMN", 1],
+]);
+
+export const convertToCategoryList = (
+  ruleList: RuleTemplate[]
+): RuleCategory[] => {
+  const dict = ruleList.reduce((dict, rule) => {
+    if (!dict[rule.category]) {
+      const id = rule.category.toLowerCase();
+      dict[rule.category] = {
+        id: rule.category,
+        ruleList: [],
+      };
+    }
+    dict[rule.category].ruleList.push(rule);
+    return dict;
+  }, {} as { [key: string]: RuleCategory });
+
+  return Object.values(dict).sort(
+    (c1, c2) =>
+      (categoryOrder.get(c2.id as CategoryType) || 0) -
+      (categoryOrder.get(c1.id as CategoryType) || 0)
+  );
+};
+
+// TODO: i18n
+export const ruleList: RuleTemplate[] = [
+  {
+    id: "engine.mysql.use-innodb",
+    category: "ENGINE",
+    engine: "MYSQL",
+    description: "Require InnoDB as the storage engine.",
+    level: RuleLevel.ERROR,
```

Comment:
level doesn't seem to belong here since it's a runtime attribute configured by the user. Field in the template should be pre-determined and can not be changed by the user.

Reply:
I think we should need it as the default error level. Otherwise, we have to define the interface for the rule, the rule in the template, and the rule in API. Back to the `Rule` and the `SelectedRule` before.
Currently, the `RuleTemplate` is for all the rule related UX, and will convert to `SchemaPolicyRule` for the API call.